### PR TITLE
Sort submissions by status relevance

### PR DESCRIPTION
**Pull request by ghost**
Addresses firecloud-app #105. Instead of sorting alphabetically by job status, sort by the weighted relevance number assigned in the functions called by `sort-order-submissions` which match to the icons that are displayed. 

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [ ] **LR** sign off
- [x] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by davidangb** [comment]
How about refactoring this so we don't have the business logic copied in two places? I am picturing:
* one set of functions (for submissions and workflows) that returns a number. We'll use that number for sorting
* a function that translates this number into an icon
* another set of convenience functions (for submissions and workflows) that returns an icon

it could look something like the following - fair warning, following is untested!!!!

```
(defn sort-order-sub-status [wf-statuses]
  (cond
    (contains? wf-statuses :Failed) 20
    (contains? wf-statuses :Aborted) 30
    (contains? wf-statuses :Succeeded) 40
    :else (do (utils/log "Unknown submission status: " wf-statuses) 50))
  )

(defn sort-order-wf-status [status]
  (cond
    (contains? wf-running-statuses status) 10
    (contains? wf-failure-statuses status) 30
    (contains? wf-success-statuses status) 40
    :else (do (utils/log "Unknown workflow status: " status) 50)))

(defn sort-order-submission [sub-status wf-statuses]
  (if (= "Done" sub-status)
    (sort-order-sub-status wf-statuses)
    (sort-order-wf-status sub-status)))

(defn- icon-for-sort-index [sort-index]
  (case sort-index
    10 (render-running-icon)
    20 (render-failure-icon)
    30 (render-failure-icon)
    40 (render-success-icon)
    50 (render-unknown-icon)
    (render-unknown-icon)))

(defn icon-for-wf-status [status]
  (icon-for-sort-index (sort-order-wf-status status)))

(defn icon-for-sub-status [wf-statuses]
  (icon-for-sort-index (sort-order-sub-status wf-statuses)))

(defn icon-for-submission [sub-status wf-statuses]
  (icon-for-sort-index (sort-order-submission sub-status wf-statuses)))
```

**Comment by ghost** [comment]
I like this approach!

**Comment by davidangb** [comment]
I like the refactor! Two more comments:
1. in functional testing, this sorts "Done" submissions separate from each other, which is a change from current:
fireshot capture 29 - firecloud i broad institute_ - http___local broadinstitute org_ w
Probably this needs to be changed; might need PO input on this.

2. Not a blocker for this PR, but always nice and could be a good learning exercise: these functions could be unit-tested! See [TESTING.md](TESTING.md).

File: src/cljs/test/broadfcuitest/sort_order_submission.cljs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                  {:status "Done" :workflowStatuses {:Aborted 2 :Succeeded 3}}
+                  {:status "Done" :workflowStatuses {:Aborted 2 :Succeeded 3}}
+                  {:status "Done" :workflowStatuses {:Aborted 2 :Succeeded 3}}
+                  {:status "Done" :workflowStatuses {:Succeeded 3}}
+                  {:status "Submitted"}
+                  {:status "Submitted"}
+                  {:status "Aborting" :workflowStatuses {:blah 3}}
+                  {:status "Aborted" :workflowStatuses {:Aborted 3}}
+                  {:status "Aborting" :workflowStatuses {:blah 3}}
+                  {:status "Aborted" :workflowStatuses {:Aborted 1 :Failed 2 :Succeeded 3}}]]
```
Comment (davidangb) [inline src/cljs/test/broadfcuitest/sort_order_submission.cljs]: could you add to this mixed test:
1. a submission status of "Blah"
3. one or more submissions of status "Succeeded"
4. a submission status of "Done" but no workflows
5. a submission status of "Done" and `:workflowStatuses {:blah 3}`

in other words - I don't think this test exercises codes 50, 80, or 90.

File: src/cljs/main/broadfcui/page/workspace/monitor/common.cljs [hunk 2]
```diff
@@ -81,27 +81,47 @@
...
+  (case sort-index
+    20 (render-failure-icon)
+    30 (render-failure-icon)
+    40 (render-success-icon)
+    60 (render-running-icon)
+    70 (render-failure-icon)
+    50 (render-unknown-icon)
+    80 (render-success-icon)
+    90 (render-unknown-icon)
+    (render-unknown-icon)))
```
Comment (davidangb) [inline src/cljs/main/broadfcui/page/workspace/monitor/common.cljs]: in order to understand the code and do a review, I wrote out comments for this block:
```
(defn- icon-for-sort-index [sort-index]
  (case sort-index           ; icon / submission status / workflow status
    20 (render-failure-icon) ; ! / Done / at least one Failed
    30 (render-failure-icon) ; ! / Done / at least one Aborted
    40 (render-success-icon) ; ✓ / Done / at least one Succeeded
    50 (render-unknown-icon) ; ? / unknown submission status / any
    60 (render-running-icon) ; = / Running | Submitted | Queued | Launching / any
    70 (render-failure-icon) ; ! / Failed | Aborting | Aborted / any
    80 (render-success-icon) ; ✓ / Succeeded / any
    90 (render-unknown-icon) ; ? / any except Done / no workflows, or all workflows are unknown
    (render-unknown-icon)))  ; ? / something went wrong in the code
```
is this correct? If so, include or abandon at your discretion!

**Review by davidangb — APPROVED** [review]
one small test improvement inline, otherwise :+1:

File: src/cljs/main/broadfcui/page/workspace/monitor/common.cljs (hunk 2 above)
Comment (ghost) [reply src/cljs/main/broadfcui/page/workspace/monitor/common.cljs]: it's close to correct! i think it's a good thing to have, will adjust and keep.

File: src/cljs/test/broadfcuitest/sort_order_submission.cljs (hunk 1 above)
Comment (ghost) [reply src/cljs/test/broadfcuitest/sort_order_submission.cljs]: Adding suggestions 1), 3) and 4). Regarding 2), based on code here: [`ExecutionModel.SubmissionStatuses`](https://github.com/broadinstitute/rawls/blob/c25c8736f96fc020430958a641051bfeda805d54/core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala#L454) I don't think we'll see a submission status of "Succeeded", so I don't think it makes sense to add it as a test case.

**Comment by jbuabbud** [comment]
@irosenbe do you approve of this sort order? If not please let me know, I'm happy to revise. 
screen shot 2018-07-26 at 1 06 08 pm

**Comment by ghost** [comment]
Finalized sorting: 
screen shot 2018-07-30 at 4 27 40 pm

**Review by davidangb — APPROVED** [review]
:+1: again! (did not test)

**Comment by ghost** [comment]
seeing unrelated swat tests failures - jenkins retest

**Comment by ghost** [comment]
unrelated known swat test failures - jenkins retest

**Pull request merged**